main.py

```diff
@@ -0,0 +1,341 @@
+    SerializedDataCoordinate,
+    SerializedDatasetType,
+    SerializedDatasetRef,
+    SerializedDimensionRecord,
+    DatasetRef,
+    DimensionConfig,
+)
+from lsst.daf.butler.core.serverModels import (
+    ExpressionQueryParameter,
+    QueryDatasetsModel,
+    QueryDataIdsModel,
+    QueryDimensionRecordsModel,
+)
+from lsst.daf.butler.core.utils import globToRegex
+from lsst.daf.butler.registry import CollectionType, CollectionSearch
+
+BUTLER_ROOT = "/Volumes/MediaHD/Users/timj/work/lsstsw/build/ci_hsc_gen3/DATA"
+
+COLLECTION_TYPE_RE = "^(" + "|".join(t.name for t in CollectionType.all()) + ")$"
+
+log = logging.getLogger("excalibur")
+
+
+class MaximalDataId(BaseModel):
+    """Something that looks like a DataId but isn't."""
+    detector: Optional[int] = None
+    instrument: Optional[str] = None
+    physical_filter: Optional[str] = None
+    exposure: Optional[int] = None
+
+
+class BindParams(BaseModel):
+    """Container for generic dict."""
+    data: Dict[str, Any]
+
+
+app = FastAPI()
+app.add_middleware(GZipMiddleware, minimum_size=1000)
+
+
+@app.get("/")
+def read_root():
+    return "Welcome to Excalibur... aka your Butler Server"
+
+
+@app.get("/butler.json")
+def read_server_config() -> Mapping:
+    """Return the butler configuration that the client should use."""
+    config_str = f"""
+datastore:
+    root: {BUTLER_ROOT}
+registry:
+    cls: lsst.daf.butler.registry.RemoteRegistry
+    db: <butlerRoot>
+"""
+    config = Config.fromString(config_str, format="yaml")
+    return config
+
+
+@app.get("/universe")
+def get_dimension_universe() -> DimensionConfig:
+    """Allow remote client to get dimensions definition."""
+    butler = Butler(BUTLER_ROOT)
+    return butler.registry.dimensions.dimensionConfig
+
+
+@app.get("/uri/{id}")
+def get_uri(id: Union[int, UUID]) -> str:
+    """Return a single URI of non-disassembled dataset."""
+    butler = Butler(BUTLER_ROOT)
+    ref = butler.registry.getDataset(id)
+    uri = butler.datastore.getURI(ref)
+
+    # In reality would have to convert this to a signed URL
+    return str(uri)
+
+
+@app.get(
+    "/registry/datasetType/{datasetTypeName}",
+    summary="Retrieve this dataset type definition.",
+    response_model=SerializedDatasetType,
+    response_model_exclude_unset=True,
+    response_model_exclude_defaults=True,
+    response_model_exclude_none=True,
+)
+def get_dataset_type(datasetTypeName: str) -> SerializedDatasetType:
+    butler = Butler(BUTLER_ROOT)
+    datasetType = butler.registry.getDatasetType(datasetTypeName)
+    return datasetType.to_simple()
+
+
+@app.get(
+    "/registry/datasetTypes",
+    summary="Retrieve all dataset type definitions.",
+    response_model=List[SerializedDatasetType],
+    response_model_exclude_unset=True,
+    response_model_exclude_defaults=True,
+    response_model_exclude_none=True,
+)
+def query_all_dataset_types(components: Optional[bool] = None) -> List[SerializedDatasetType]:
+    butler = Butler(BUTLER_ROOT)
```

Comment:
I hope we're creating a new `Butler` instance for each API call for simplicity (that'd be a good idea) rather than because it'd be a pain to have persistent state.  We try to keep `Butler` startup times pretty reasonable, but they're definitely not instantaneous.

Reply:
Yes, this was all growing as I went along experimenting. Given that we know that this server is attaching to only one registry it is going to be better to create one Butler on startup and then create a butler from that butler.

Reply:
That would let me prefetch storage classes and dataset types.

Reply:
Funnily enough, pre-loading the butler makes the client/server `butler query-collections` command line slightly faster than the non-client server one (for ci_hsc_gen3).